src/osd/PG.cc

```diff
@@ -529,11 +503,11 @@ void PG::_finish_recovery(Context *c)
     if (scrub_after_recovery) {
       dout(10) << "_finish_recovery requeueing for scrub" << dendl;
       scrub_after_recovery = false;
-      scrubber.must_deep_scrub = true;
-      scrubber.check_repair = true;
-      // We remember whether req_scrub was set when scrub_after_recovery set to true
-      scrubber.req_scrub = save_req_scrub;
-      queue_scrub();
+      m_planned_scrub.must_deep_scrub = true;
+      m_planned_scrub.check_repair = true;
+      m_planned_scrub.must_scrub = true; // RRR added this line, as I don't see how the comment in PG::sched_scrub_initial() is true otherwise
```

Comment:
Setting must_scrub isn't required in the original code because queue_scrub() knows we are scrubbing, so it doesn't matter in that function if this is a scheduled scrub or user initiated scrub.

Reply:
This is why req_scrub is added, to remember whether original scrub was user requested.  Also, the comment in sched_scrub_initial() is true at a higher level.  Internally, it doesn't have to be so strict, but maybe it is ok to set it here as long as it is cleared.  We don't want that flag on during scrubbing.  We don't want to show MUST_SCRUB generally in the log, so it is cleared once scrub starts.

Reply:
Thanks for the clarification.

(req_scrub wasn't there yet when I added the comment you ref'ed to.)

I have found the flags handling the hardest to follow, and made the following changes - mostly for structure & legibility - while following all the details of the existing implementation.

The flags are now divided into the following categories/structures:

### (A) Flags managed by the PG to affect the *next* scrub parameters (and scheduling):
Grouped in PG.m_planned_scrub of type `requested_scrub_t`.

- Primary only;
- all `next-scrub` ((next commit push later today)) flags are cleared on `PG::sched_scrub()`, after affecting the running scrub flags (B below).
Note that this is different from the existing implementation, though I think I maintained the existing behaviour.
Specifically:
  - **must_scrub**: was cleared in the original code. A copy is now maintained in **marked_must** (see below).   
  - **need_auto**, **must_deep_scrub**: cleared in the original code too;    
  - **must_repair**: same. State is remembered in PG_STATE_REPAIR.    
  - **deep_scrub_on_error**: copied to the running flags. That copy is only cleared at `scrub_finish()`, after initiating an auto deep-scrub.
  - **check_repair**: original behavior is maintained.
  - **time_for_deep**: _TBD... I am verifying the code now..._
  - **auto_repair**: state affects running scrub 'auto_repair' flag; also saved in PG_STATE_REPAIR. Clearing matches original code. 
  - **req_scrub** - note that the new version uses `m_planned_scrub.req_scrub` for the role played by save_req_scrub:
when `scrub_finish()` determines that a scrub_after_recovery is in order, it copies ('or'ing, actually) the current
**required** flag (what was **req_scrub** when we started) back into the next-scrub's flag in `m_planned_scrub`.   
- note the detailed description of each flag, per my understanding of the code, in scrub_common.h;

### (B) Flags used by the executing scrub. These are owned by the `PgScrubber`.
Grouped in `PgScrubber.m_flags`, of type `scrub_flags_t`.  
- used by both primary and replicas;
- set in `PgScrubber::set_op_parameters()` (for a Primary), which is called from:
  - `PG::sched_scrub()` - and note that we never reach here again after the scrubbing starts; and
  - `PG::queue_scrub_after_recovery()`

and by `PgScrubber::replica_scrub_op()`
- a new **marked_must** flag captures the state of **must_scrub** at start time, to affect the scrubber
periodic sleeps before each NewChunk (note that I didn't see that being handled in the original code.
Seems like it uses `must_scrub` that was already cleared in `PG::sched_scrub()`).
- cleared when scrubbing finishes, in `PgSCrubber::reset_internal_state()`.

Reply:
Please note that req_scrub stands for requested scrub which is the same as marked_must.  req_scrub was only set in limited cases where a new scrub would happen and we needed to keep track of whether the instigating scrub was user requested.  If you set req_scrub early in the scrub process, you just need to be careful not to clear it in the cases where it is needed.

Reply:
Currently there are some differences in the ways req_scrub vs mark_scrub affect scrubbing behaviour. After discussing this with others - I will merge the two anyway.